packages/edit-site/src/components/navigation-sidebar/navigation-panel/menus/template-parts.js

```diff
@@ -21,13 +21,15 @@ import { MENU_ROOT, MENU_TEMPLATE_PARTS } from '../constants';
 
 export default function TemplatePartsMenu() {
 	const templateParts = useSelect( ( select ) => {
-		return select( 'core' ).getEntityRecords(
-			'postType',
-			'wp_template_part',
-			{
+		const unfilteredTemplateParts =
+			select( 'core' ).getEntityRecords( 'postType', 'wp_template_part', {
 				status: [ 'publish', 'auto-draft' ],
 				per_page: -1,
-			}
+			} ) || [];
+		const currentTheme = select( 'core' ).getCurrentTheme()?.stylesheet;
+		return unfilteredTemplateParts.filter(
+			( item ) =>
+				item.status === 'publish' || item.meta.theme === currentTheme
```

Comment:
Not sure how solid this suggestion is, but what about fetching template parts in 2 separate requests?
E.g.
```js
getEntityRecords( 'postType, 'wp_template_part', {
  status: 'publish',
  per_page: -1
} );
getEntityRecords( 'postType, 'wp_template_part', {
  status: 'auto-draft',
  theme: currentTheme,
  per_page: -1
} );
```

Maybe using the `meta` would make things slower, but if (once?) we replace it with a taxonomy, it should be snappier I guess?

Just to be clear, this is not a blocker or anything, just an idea that occurred to me.